src/db/spanner/models.rs

```diff
@@ -0,0 +1,1302 @@
+                    println!("err {}", e);
+                    // TODO Return error
+                    Ok(0)
+                }
+                Ok(res) => {
+                    println!("ok {:?}", res.1);
+                    match res.1.rows {
+                        Some(row) => Ok(row[0][0].parse().unwrap()),
+                        None => Ok(0),
+                    }
+                }
+            };
+            if let Ok(val) = rv {
+                self.coll_cache.put(val, name.to_owned())?;
+            };
+        };
+        // let id = self.conn.transaction(|| {
+        //     sql_query("INSERT INTO collections (name) VALUES (?)")
+        //         .bind::<Text, _>(name)
+        //         .execute(&self.conn)?;
+        //     collections::table.select(last_insert_id).first(&self.conn)
+        // })?;
+        // self.coll_cache.put(id, name.to_owned())?;
+        id
+    }
+
+    fn get_or_create_collection_id(&self, name: &str) -> Result<i32> {
+        self.get_collection_id(name).or_else(|e| match e.kind() {
+            DbErrorKind::CollectionNotFound => self.create_collection(name),
+            _ => Err(e),
+        })
+    }
+
+    pub fn lock_for_read_sync(&self, params: params::LockCollection) -> Result<()> {
+        let user_id = params.user_id.legacy_id as u32;
+        let collection_id =
+            self.get_collection_id(&params.collection)
+                .or_else(|e| match e.kind() {
+                    // If the collection doesn't exist, we still want to start a
+                    // transaction so it will continue to not exist.
+                    DbErrorKind::CollectionNotFound => Ok(0),
+                    _ => Err(e),
+                })?;
+        // If we already have a read or write lock then it's safe to
+        // use it as-is.
+        if self
+            .inner
+            .session
+            .borrow()
+            .coll_locks
+            .get(&(user_id, collection_id))
+            .is_some()
+        {
+            return Ok(());
+        }
+
+        // Lock the db
+        self.begin()?;
+        let spanner = &self.conn;
+        let session = spanner.session.name.as_ref().unwrap();
+        let mut sql = ExecuteSqlRequest::default();
+        sql.sql = Some("SELECT CURRENT_TIMESTAMP() as now, last_modified FROM user_collections WHERE userid=@userid AND collection=@collectionid LOCK IN SHARE MODE".to_string());
+        let mut params = HashMap::new();
+        params.insert("userid".to_string(), user_id.to_string());
+        params.insert("collection".to_string(), collection_id.to_string());
+        sql.params = Some(params);
+
+        let results = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        if let Ok(results) = results {
+            if let Some(rows) = results.1.rows {
+                let modified = SyncTimestamp::from_i64(rows[0][0].parse().unwrap())?;
```

Comment:
So Spanner TIMESTAMP types come back as i64 (probably in the raw json)?

The Python Spanner client brings them back as Python datetimes -- probably because it's using grpc underneath w/ more type info

Reply:
We'll have to double check a lot of assumptions like this made by the code once the tests are running against spanner and exposing errors to fix.

Right now the tests run against spanner, but all fail silently. Need to figure out why no errors are being printed anywhere.